### PR TITLE
feat(auth): Expose modular API that matches the Firebase web JS SDK v9 API

### PR DESCRIPTION
**Pull request by Lyokone**
### Related issues

https://github.com/invertase/react-native-firebase/issues/6705

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [ ] Yes
- My change supports the following platforms;
  - [ ] `Android`
  - [ ] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [ ] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Review by mikehardy — COMMENTED** [review]
just to note passes all tests locally, structurally looks right, I'm running through index.js right now with an eye towards how platform differences were handled (always a point of discussion, but already discussed I think...) and why github is telling me that there was no test coverage on the new modular APIs (when it looks like they were covered... 🤔 )

**Review by mikehardy — CHANGES_REQUESTED** [review]
All seems to work, and the cause of coverage notices appears to be that the declaration lines are not covered but the implementation lines are? This seems odd and worth a (separate, not related to this PR) investigation but shouldn't be a blocker. Hypothesis based on looking at this: https://app.codecov.io/gh/invertase/react-native-firebase/pull/7015

I have a couple real questions though:

- can we have *only* the modular e2e tests, with their non-modular compat testing inside of them, and remove the non-modular tests? Really, just make the modular tests the new default, no need to have modular in the name
- the provider, rnReload, and multifactor tests do not seem to be modularized, if those are not modularized / exercising those features in the v9 / modular way I fear they will break in the future. Unless I am misunderstanding (I could be! please let me know) I think that's a blocker for merge

**Comment by Lyokone** [comment]
Thanks for the review.
- For the first point, all the other migrated modules have kept the `.modular` for the new tests. I think we can remove the non-modular of all plugins at the same time but I don't see a reason to do it only for auth.
- For the second point, I missed that you could access multifactor with a top level function now, I'll add the 3 missing e2e tests.

File: packages/auth/e2e/rnReload.modular.e2e.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        currentUser.uid.should.be.a.String();
+        currentUser.toJSON().should.be.an.Object();
+        should.equal(currentUser.toJSON().email, null);
+        currentUser.isAnonymous.should.equal(true);
+        currentUser.providerId.should.equal('firebase');
+        currentUser.should.equal(defaultAuth.currentUser);
+
+        // test correct user is returned after signing
+        // in with a different user then reloading
+        await signOut(defaultAuth);
```
Comment (Lyokone) [inline packages/auth/e2e/rnReload.modular.e2e.js:162]: This signOut hangs out, I have no idea why @mikehardy

Comment (mikehardy) [reply packages/auth/e2e/rnReload.modular.e2e.js:162]: Huh - that is unexpected (of course!) - I'll check it locally and see what I can do about CI / investigating

**Comment by mikehardy** [comment]
> For the first point, all the other migrated modules have kept the .modular for the new tests. I think we can remove the non-modular of all plugins at the same time but I don't see a reason to do it only for auth.


Okay - agreed there, I think we did one of them (the most recent?) with just convert-main-file-name-to-modular + compat-inside-modular and I'd like that going forward. You are correct that since there are quite a few in the old style (that this one is using) of old-style + modular-with-compat-inside we can do one PR that realigns all at once. No need to hold this up waiting for that

I'll check the new tests and see what I can find out about signOut hanging. Thanks @Lyokone !

**Review by russellwheatley — CHANGES_REQUESTED** [review]
I think a PhoneAuthProvider test needs to be updated. And there's the lint failure that Mike mentioned. Other than that, LGTM 👌.

**Comment by mikehardy** [comment]
I fixed the PhoneAuthProvider constructor to throw in a  backwards-compatible way and cleared the lint error, rebased to current main and re-pushed, we'll see what CI says, seems to run fine locally...

**Comment by Lyokone** [comment]
@mikehardy Thanks for taking the on the rest of the PR 👍

**Comment by mikehardy** [comment]
Note to self: need to add API for #7148 that came in while this was in-flight

**Comment by mikehardy** [comment]
Please do not add merge commits unless necessary for some reason, they make it more difficult for me to reason about PRs
In this case I don't think any new commits were added but main/HEAD was merged anyway, which does not seem to me to add any value / serve any purpose but does make the commit string unclean

**Comment by mikehardy** [comment]
I think everything is fixed now, pending CI

**Comment by mikehardy** [comment]
Passed all the tests except lint on the last run, I fixed the lint one - merging!

**Comment by aedocoding** [comment]
Hello, I was wondering if this PR being merged means that the next release of `react-native-firebase` would have the [Config interface](https://firebase.google.com/docs/reference/js/auth.config) accessible to users to make changes to properties like the `apiHost` or `tokenApiHost`? Apologies if this is the wrong place to ask this question by the way

**Pull request merged**